Avoid adding drops for types w/ no dtor in MIR construction

wesleywiser opened the pull request:
Fixes #28159

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

wesleywiser commented:
I'm not really sure how to write an automated test for this. If someone can point me in the right direction, I'd be happy to add one.

nagisa commented on src/librustc/middle/ty/mod.rs:
```diff
@@ -2396,6 +2396,45 @@ impl<'tcx> ctxt<'tcx> {
+    /// (Note that this implies that if `ty` has a destructor attached,
+    /// then `type_needs_drop` will definitely return `true` for `ty`.)
+    pub fn type_needs_drop(&self, ty: Ty<'tcx>) -> bool {
+        self.type_needs_drop_given_env(ty, &self.empty_parameter_environment())
```
Is it really fine to use empty param environment here? I assume this is supposed to hold some sort of substs for generics. It was fine to do pass empty environment in trans, because all types are monomorphized there, but in `middle::ty` or `mir::hair::cx` you cannot assume that types will be monomorphized (they usually won’t be).

pnkfelix commented on src/librustc/middle/ty/mod.rs:
```diff
@@ -2396,6 +2396,45 @@ impl<'tcx> ctxt<'tcx> {
+                                         ty: Ty<'tcx>,
+                                         param_env: &ty::ParameterEnvironment<'a,'tcx>) -> bool {
+        // Issue #22536: We first query type_moves_by_default.  It sees a
+        // normalized version of the type, and therefore will definitely
```
The same note that @nagisa wrote on `fn type_needs_drop` applies here, namely, that the assumptions that hold in the context of `trans` may not hold here. (When I wrote "normalized" in the original comment, I probably should have written "monomorphized.")

nagisa commented:
Interestingly we already have [a check](https://github.com/rust-lang/rust/blob/42c3ef8f9fd4b0dd1f881c49323bad456163f202/src/librustc_mir/build/scope.rs#L251) so perhaps adjusting [this](https://github.com/rust-lang/rust/blob/42c3ef8f9fd4b0dd1f881c49323bad456163f202/src/librustc_mir/hair/cx/mod.rs#L106) would be better?

nagisa commented on src/librustc_mir/hair/cx/mod.rs:
```diff
@@ -105,11 +105,7 @@ impl<'a,'tcx:'a> Cx<'a, 'tcx> {
-            // (`type_needs_drop`) is currently factored into
-            // `librustc_trans`, so we can't easily do so.
-            true
+            self.tcx.type_needs_drop(ty)
```
Note, that here infer context is available, and thus it should possible to monomorphise the type (see/copy `rustc::middle::infer::InferCtxt::expr_ty_adjusted`).

wesleywiser replied:
Shouldn't I just pass in `self.infcx.parameter_environment` here?

nagisa replied:
That could also work.

wesleywiser commented:
@nagisa @pnkfelix I've updated the code to pass along `self.infcx.parameter_environment`. From your feedback, I don't think it makes sense to have `type_needs_drop` in `middle` since any code at that point should be passing in a `ParameterEnvironment` so I've moved `type_needs_drop` back into `trans`.

pnkfelix commented on src/librustc_mir/hair/cx/mod.rs:
```diff
@@ -105,11 +105,7 @@ impl<'a,'tcx:'a> Cx<'a, 'tcx> {
 
     pub fn needs_drop(&mut self, ty: Ty<'tcx>, span: Span) -> bool {
         if self.infcx.type_moves_by_default(ty, span) {
```
Is this first `if` redundant now?  Maybe the whole method body can be simplified  to a single function call?

pnkfelix commented:
Seems fine apart from the potential redundancy noted above. r=me once that is either removed or explained (in the latter case,  a comment on this PR would suffice)

wesleywiser commented:
`type_moves_by_default` simply checks if the type implements `Copy`. `type_needs_drop_given_env` already checks that so I think you are correct and the function can be simplified.

pnkfelix commented:
@bors r+ 8c897ee

bors commented:
:hourglass: Testing commit 8c897ee with merge dc1f442...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7577)

pnkfelix commented:
@bors retry

bors commented:
:zap: Previous build results for [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/875), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/786), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2738), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7629), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7622), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7587), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2608), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1839) are reusable. Rebuilding only [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7601), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7620), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7558), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7577), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7561), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2592), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2606), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2587), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2588)...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7603), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7622), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/875), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7560), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7579), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7563), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/786), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2738), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7629), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7622), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7587), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2594), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2608), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2608), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2589), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1839), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2590)

wesleywiser commented:
Thanks everybody!

On Wed, Jan 6, 2016, 2:29 AM bors <notifications@github.com> wrote:

> Merged #30492 <https://github.com/rust-lang/rust/pull/30492>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/30492#event-506792318>.
>
